docs/design/RUC-on-type.md

```diff
@@ -0,0 +1,189 @@
+
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode
+{
+    class NestedClass {
+        // ILXXXX: The method will generate a warning that will not be suppressed
+        public static void MethodWithDangerousCode ()
+        { 
+            // Dangerous code that generetares a trimming warning
+        }
+    }
+}
+
+static void TestSuppressionOnType ()
+{
+    ClassWithRequiresUnreferencedCode.NestedClass.MethodWithDangerousCode()
+}
+```
+
+## Derived types
+There is a special handling for types deriving from a base type which is annotated with RequiresUnreferencedCode. Although we dont auto propagate the RequiresUnreferencedCode attribute to the derived type a warning is generated in the derived type declaration to warn the user that calling methods in the derived type might end up calling something dangerous at it's base. Therefore, is recomendable to also mark the derived type with RequiresUnreferencedCode.
+
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode
+{
+    public static void StaticMethod () { }
+}
+
+// IL2109: Type 'DerivedWithoutRequires' derives from 'ClassWithRequiresUnreferencedCode' which has 'RequiresUnreferencedCodeAttribute'.  Message for --ClassWithRequiresUnreferencedCode--.
+class DerivedWithoutRequires : ClassWithRequiresUnreferencedCode
+{
+    public static void StaticMethodInInheritedClass () { }
+}
+
+// TestRequiresOnBaseButNotOnDerived(): Using method 'ClassWithRequiresUnreferencedCode.StaticMethod()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Message for --ClassWithRequiresUnreferencedCode--.
+static void TestRequiresOnBaseButNotOnDerived ()
+{
+    DerivedWithoutRequires.StaticMethodInInheritedClass ();
+    DerivedWithoutRequires.StaticMethod ();
+}
+```
+As the previous example ilustrates the DerivedType without RequiresUnreferencedCode will generate a warning only for deriving from ClassWithRequiresUnreferencedCode. Additionally, by calling some of the methods we can see the difference on behavior on methods declared on the derived type vs methods on the base type. Calling a method in the derived type will not generate a warning since the derived type is not annotated, while calling a method for which the base type will be called will generate IL2026.
+
+Arguably we could say that the warning in the type declaration is not necessary and that we could rely on the warning being generated in the constructor instead, but this could lead to some behaviors in which we can call a method without ever using the constructor.
+
+```C#
+var t = Type.GetType("Derived").GetMethod("SomeMethod").Invoke(null, null);
+
+[RequiresUnreferencedCode("This base type is dangerous")]
+class Base {}
+
+class Derived : Base 
+{
+    static void SomeMethod() {}
+}
```

Comment:
Is this example really problematic?

Derived.SomeMethod is not marked as RUC because we didn't propagate RUC from Base. So no warnings are suppressed in it. If it didn't generate any warnings, it will be safe to reflection-invoke. If it generated warnings, it just means that the class author has to go and either selectively annotate SomeMethod as RUC, or RUC annotate the whole type. Then the reflection will warn too.